admin-notice/class-admin-notice-controller.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Automattic\VIP\Admin_Notice;
+
+class Admin_Notice_Controller {
+	private $all_notices = [];
+
+	public function init() {
+		add_action( 'admin_notices', [ $this, 'display_notices' ] );
+	}
+
+	public function add( Admin_Notice $notice ) {
+		array_push( $this->all_notices, $notice );
+	}
+
+	public function display_notices() {
+		$filtered_notices = array_filter( $this->all_notices, function ( $notice ) {
+			return $notice->should_render();
+		});
+
+		$html_notices = array_map( function ( $notice ) {
+			return $notice->to_html();
+		}, $filtered_notices );
+
+		foreach ( $html_notices as $html_notice ) {
+			echo( wp_kses_post( $html_notice ) . "\n" );
```

Comment:
The presence of the WordPress escaping _here_ bugs me.

Rather than having this controller tied to the escaping, consider doing:

```php
public function display_notices() {
	// Get $filtered_notices as currently.

	foreach( $filtered_notices as $notice ) {
		$notice->display();
	}
}
```

That way, we can handle all of the escaping as the HTML markup is being generated and displayed, from within the Admin_Notice itself (`display()` method), so that any class that extends Admin_Notice can adjust it's escaping as needed (perhaps to allow `wp_kses()` with a custom set of elements -> attributes that allow for particular aria-* attributes).

Reply:
I am still not sold on this as I don't like the side effect of printing to STDOUT. buut you have good points as well and I don't have strong opinion on this one. So I changed it as you suggested.